File: app/src/main/java/im/fdx/v2ex/ui/LoginActivity.java
package im.fdx.v2ex.ui;

import android.app.ProgressDialog;
import android.content.Intent;
import android.content.SharedPreferences;
import android.net.Uri;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.design.widget.TextInputEditText;
import android.support.v4.content.LocalBroadcastManager;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.text.TextUtils;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;

import com.elvishew.xlog.XLog;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

import java.io.IOException;

import im.fdx.v2ex.MyApp;
import im.fdx.v2ex.R;
import im.fdx.v2ex.model.MemberModel;
import im.fdx.v2ex.network.HttpHelper;
import im.fdx.v2ex.network.NetManager;
import im.fdx.v2ex.utils.HintUI;
import im.fdx.v2ex.utils.Keys;
import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.FormBody;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

import static android.util.Base64.encodeToString;
import static im.fdx.v2ex.network.NetManager.API_USER;
import static im.fdx.v2ex.network.NetManager.HTTPS_V2EX_BASE;
import static im.fdx.v2ex.network.NetManager.SIGN_IN_URL;
import static im.fdx.v2ex.network.NetManager.myGson;

public class LoginActivity extends AppCompatActivity implements View.OnClickListener {
    private static final String TAG = LoginActivity.class.getSimpleName();

    public static final int LOG_IN_SUCCEED = 2;
    private static final int LOG_IN_FAILED = 3;
    public static final String ANDROID_KEY_STORE = "AndroidKeyStore";
    private TextInputEditText etUsername;
    private TextInputEditText etPassword;
    private ProgressDialog progressDialog;
    private SharedPreferences mSharedPreference;
    private String username;
    private String password;
    String avatar;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_login);

        mSharedPreference = PreferenceManager.getDefaultSharedPreferences(this);



        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        if (toolbar != null) {
            setSupportActionBar(toolbar);
            ActionBar actionBar = getSupportActionBar();
            toolbar.setTitle("");
            if (actionBar != null) {
                actionBar.setDisplayHomeAsUpEnabled(true);
            }
            toolbar.setNavigationOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    onBackPressed();
                }
            });
        }

        etUsername = (TextInputEditText) findViewById(R.id.input_username);
        progressDialog = new ProgressDialog(this, R.style.AppTheme_Light_Dialog);
        String usernamePref = mSharedPreference.getString("username", "");
        etPassword = (TextInputEditText) findViewById(R.id.input_password);

        Button btnLogin = (Button) findViewById(R.id.btn_login);
        TextView tvSignup = (TextView) findViewById(R.id.link_sign_up);


        btnLogin.setOnClickListener(this);
        tvSignup.setOnClickListener(this);

        if (!TextUtils.isEmpty(usernamePref)) {
            etUsername.setText(usernamePref);
            etPassword.requestFocus();
        }

    }


    @Override
    protected void onResume() {
        super.onResume();
//        login();
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.btn_login:
                if (!isValidated()) return;
                login();
                break;
            case R.id.link_sign_up:
                Intent openUrl = new Intent(Intent.ACTION_VIEW, Uri.parse(NetManager.SIGN_UP_URL));
                startActivity(openUrl);
                break;
        }
    }


    private void login() {

        progressDialog.setIndeterminate(true);
        progressDialog.setMessage(getString(R.string.authenticating));
        progressDialog.show();
        username = etUsername.getText().toString();
        password = etPassword.getText().toString();

        getLoginData();


    }

    private void getLoginData() {
        Request requestToGetOnce = new Request.Builder().headers(HttpHelper.baseHeaders)
                .url(SIGN_IN_URL)
                .build();

        HttpHelper.OK_CLIENT.newCall(requestToGetOnce).enqueue(new Callback() {
            @Override
            public void onFailure(Call call, IOException e) {
                Log.e("FDX", "error in get login page");
            }

            @Override
            public void onResponse(Call call, Response response0) throws IOException {
                String htmlString = response0.body().string();
                Element body = Jsoup.parse(htmlString).body();
//                    Element box = body.getElementsByClass("box").last();
//                    Element cell = box.getElementsByClass("cell").first();
                String nameKey = body.getElementsByAttributeValue("placeholder", "用户名或电子邮箱地址").attr("name");
                String passwordKey = body.getElementsByAttributeValue("type", "password").attr("name");
                String onceCode = body.getElementsByAttributeValue("name", "once").attr("value");

                XLog.tag("LoginActivity").d(nameKey + "|" + passwordKey + "|" + onceCode);
//            XLog.d(request.headers().toString());
//            XLog.d(request.toString());
//            XLog.d(request.isHttps() + request.method() +request.body().toString() );

                postLogin(nameKey, passwordKey, onceCode);

            }
        });


//            XLog.d( String.valueOf(htmlString.length()));
//            XLog.d(htmlString);

    }

    private void postLogin(String nameKey, final String passwordKey, String onceCode) {
        RequestBody requestBody = new FormBody.Builder()
//                            .add("next", "/")
                .add(nameKey, username)
                .add(passwordKey, password)
                .add("once", onceCode)
                .build();

        final Request request = new Request.Builder().headers(HttpHelper.baseHeaders)
                .url(SIGN_IN_URL)
                .header("Origin", HTTPS_V2EX_BASE)
                .header("Referer", SIGN_IN_URL)
                .header("Content-Type", "application/x-www-form-urlencoded")
                .post(requestBody)
                .build();
        HttpHelper.OK_CLIENT.newCall(request).enqueue(new Callback() {


            @Override
            public void onFailure(Call call, IOException e) {
                Log.e("FDX", "error in Post Login");
            }

            @Override
            public void onResponse(Call call, Response response) throws IOException {


                final int httpcode = response.code();
                final String errorMsg = getErrorMsg(response.body().string());
                XLog.tag("LoginActivity").d("http code: " + response.code());
                XLog.tag("LoginActivity").d("errorMsg" + errorMsg);

                switch (httpcode) {
                    case 302:

                        MyApp.getInstance().setLogin(true);
                        mSharedPreference.edit()
                                .putString("username", username)
                                .putBoolean("is_login", true)
                                .apply();
//                        LoginActivity.this.setResult(LOG_IN_SUCCEED);


                        goMyHomePage();

                        runOnUiThread(new Runnable() {
                            @Override
                            public void run() {
                                progressDialog.dismiss();
                                HintUI.T(LoginActivity.this, "登录成功");
                            }
                        });

                        finish();
                        break;
                    case 200:
//                        LoginActivity.this.setResult(LOG_IN_FAILED);
                        runOnUiThread(new Runnable() {
                            @Override
                            public void run() {
                                progressDialog.dismiss();
                                HintUI.T(LoginActivity.this, "登录失败\n " + errorMsg);
                            }
                        });
                        break;
                    default:

                        break;
                }

            }
        });
    }

    private void goMyHomePage() {
        HttpHelper.OK_CLIENT.newCall(new Request.Builder()
                .headers(HttpHelper.baseHeaders)
                .url(API_USER + "?username=" + username).build()).enqueue(new Callback() {
            @Override
            public void onFailure(Call call, IOException e) {

            }

            @Override
            public void onResponse(Call call, Response response) throws IOException {

                if (response.code() != 200) {
                    return;
                }
                String body = response.body().string();
                MemberModel member = myGson.fromJson(body, MemberModel.class);
                avatar = member.getAvatarLargeUrl();

                mSharedPreference.edit().putString("avatar", avatar).apply();
                Intent intent = new Intent(Keys.ACTION_LOGIN);
                intent.putExtra(Keys.KEY_USERNAME, username);
                intent.putExtra(Keys.KEY_AVATAR, avatar);
                LocalBroadcastManager.getInstance(LoginActivity.this)
                        .sendBroadcast(intent);
            }
        });
    }


    private String getErrorMsg(String body) {

        XLog.tag(TAG).d(body);
        Element element = Jsoup.parse(body).body();
        Elements message = element.getElementsByClass("problem");
        if (message == null) {
            return "";
        }

        return message.text().trim();
    }

    private boolean isValidated() {
        boolean valid = true;
        String username = etUsername.getText().toString();
        String password = etPassword.getText().toString();

        if (password.isEmpty()) {
            etPassword.setError("密码不能为空");
            etPassword.requestFocus();
            valid = false;
        }
        if (username.isEmpty()) {
            etUsername.setError("名字不能为空");
            etUsername.requestFocus();
            valid = false;
        }
        return valid;
    }


}

File: app/src/main/java/im/fdx/v2ex/ui/LoginActivity.kt
package im.fdx.v2ex.ui

import android.annotation.SuppressLint
import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.widget.EditText
import androidx.activity.compose.setContent
import androidx.appcompat.app.AlertDialog
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.core.os.bundleOf
//import im.fdx.v2ex.R // Removed R usage for layout
import im.fdx.v2ex.R
import im.fdx.v2ex.network.*
import im.fdx.v2ex.network.NetManager.HTTPS_V2EX_BASE
import im.fdx.v2ex.network.NetManager.SIGN_IN_URL
import im.fdx.v2ex.pref
import im.fdx.v2ex.setLogin
import im.fdx.v2ex.ui.login.LoginScreen
import im.fdx.v2ex.ui.theme.V2ExTheme
import im.fdx.v2ex.utils.Keys
import im.fdx.v2ex.utils.extensions.*
import im.fdx.v2ex.view.CustomChrome
import okhttp3.*
import okhttp3.HttpUrl.Companion.toHttpUrlOrNull
import org.jsoup.Jsoup
import java.io.IOException
import com.bumptech.glide.load.model.GlideUrl
import com.bumptech.glide.load.model.LazyHeaders
import com.elvishew.xlog.XLog
import com.google.android.material.snackbar.Snackbar
import com.google.firebase.Firebase
import com.google.firebase.remoteconfig.remoteConfig

class LoginActivity : BaseActivity() {

    private var loginName: String? = null
    // View state
    private var username by mutableStateOf("")
    private var password by mutableStateOf("")
    private var captcha by mutableStateOf("")
    private var captchaUrl by mutableStateOf<GlideUrl?>(null)
    private var isLoading by mutableStateOf(false)

    // Hidden form fields
    var onceCode: String? = null
    var passwordKey: String? = null
    var nameKey: String? = null
    var imageCodeKey: String? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val toolbar = setUpToolbar() // BaseActivity setup usually, might need adjustment if using Compose Scaffold fully. 
        // For now, keep BaseActivity structure but replace content.
        // Actually setUpToolbar might depend on layout. BaseActivity source wasn't fully shown but assuming it sets up ActionBar.
        // If content is Compose, standard ActionBar might sit on top.

        setContent {
            V2ExTheme {
                 LoginScreen(
                     username = username,
                     onUsernameChange = { username = it },
                     password = password,
                     onPasswordChange = { password = it },
                     captcha = captcha,
                     onCaptchaChange = { captcha = it },
                     captchaUrl = captchaUrl,
                     isLoading = isLoading,
                     onLoginClick = {
                         if (isValidated()) {
                             performLogin()
                         }
                     },
                     onCaptchaClick = { getLoginElement() },
                     onSignUpClick = {
                         CustomChrome(this@LoginActivity).load(NetManager.SIGN_UP_URL)
                     }
                 )
            }
        }

        val usernamePref = pref.getString(Keys.KEY_USERNAME, "")
        if (!usernamePref.isNullOrEmpty()) {
            username = usernamePref!!
        }

        getLoginElement()
    }

    @Deprecated("Deprecated in Java")
    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == 144 && resultCode == RESULT_OK) {
            setLogin(true)
            finish()
        }
    }

    private fun getLoginElement() {
        val requestToGetOnce = Request.Builder()
            .url(SIGN_IN_URL)
            .build()

        HttpHelper.OK_CLIENT.newCall(requestToGetOnce).enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                loge("error in get login page")
            }

            @Throws(IOException::class)
            override fun onResponse(call: Call, response0: Response) {
                val htmlString = response0.body.string()
                val body = Jsoup.parse(htmlString).body()
                nameKey = body.getElementsByAttributeValue("placeholder", "用户名或电子邮件地址").attr("name")
                passwordKey = body.getElementsByAttributeValue("type", "password").attr("name")
                onceCode = body.getElementsByAttributeValue("name", "once").attr("value")
                imageCodeKey = body.getElementsByAttributeValueContaining("placeholder", "请输入上图中的验证码").attr("name")
                
                runOnUiThread {
                    val str = "https://www.v2ex.com/_captcha?once=$onceCode"

                    val headers: LazyHeaders = LazyHeaders.Builder()
                        .addHeader("Accept", "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8")
                        .addHeader("Accept-Charset", "utf-8, iso-8859-1, utf-16, *;q=0.7")
                        .addHeader("Accept-Language", "zh-CN,zh;q=0.8,en;q=0.6")
                        .addHeader("Host", "www.v2ex.com")
                        .addHeader("Cache-Control", "max-age=0")
                        .addHeader(
                            "User-Agent",
                            "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36" + " (KHTML, like Gecko) Chrome/58.0.3013.3 Safari/537.36"
                        )
                        .addHeader(
                            "cookie",
                            HttpHelper.myCookieJar.loadForRequest(str.toHttpUrlOrNull()!!).joinToString(separator = ";")
                        )
                        .build()
                    captchaUrl = GlideUrl(str, headers)
                }
                XLog.tag("LoginActivity").d("$nameKey|$passwordKey|$onceCode|$imageCodeKey")
            }
        })
    }

    private fun performLogin() {
        isLoading = true
        postLogin(
            nameKey ?: "", passwordKey ?: "", onceCode = onceCode
                ?: "", imageCodeKey = imageCodeKey ?: ""
        )
    }

    private fun postLogin(nameKey: String, passwordKey: String, onceCode: String, imageCodeKey: String) {
        val requestBody = FormBody.Builder()
            .add(nameKey, username)
            .add(passwordKey, password)
            .add(imageCodeKey, captcha)
            .add("once", onceCode)
            .build()

        val request = Request.Builder()
            .url(SIGN_IN_URL)
            .header("Origin", HTTPS_V2EX_BASE)
            .header("Referer", SIGN_IN_URL)
            .header("Content-Type", "application/x-www-form-urlencoded")
            .post(requestBody)
            .build()
        HttpHelper.OK_CLIENT.newCall(request).start(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                Log.e("LoginActivity", "error in Post Login")
                runOnUiThread { isLoading = false }
            }

            @Throws(IOException::class)
            override fun onResponse(call: Call, response: Response) {
                val httpcode = response.code
                val errorMsg = Parser(response.body!!.string()).getErrorMsg()
                logd("http code: ${response.code}")
                logd("errorMsg: $errorMsg")

                when (httpcode) {
                    302 -> {
                        vCall(HTTPS_V2EX_BASE).start(object : Callback {
                            override fun onFailure(call: Call, e: IOException) {
                                e.printStackTrace()
                                runOnUiThread { isLoading = false }
                            }

                            @Throws(IOException::class)
                            override fun onResponse(call: Call, response2: Response) {
                                if (response2.code == 302) {
                                    if (("/2fa" == response2.header("Location"))) {
                                        runOnUiThread {
                                            isLoading = false
                                            showTwoStepDialog(this@LoginActivity)
                                        }
                                    }
                                } else {
                                    setLogin(true)
                                    runOnUiThread {
                                        toast("登录成功")
                                        finish()
                                    }
                                }
                            }
                        })

                    }
                    200 -> runOnUiThread {
                        // showHint(binding.root, errorMsg, Snackbar.LENGTH_INDEFINITE) // Can't use binding.root. Use Toast for now.
                        toast(errorMsg)
                        getLoginElement()
                        isLoading = false
                    }
                }

            }
        })
    }

    private fun isValidated(): Boolean {
        if (username.isEmpty()) {
            toast("名字不能为空")
            return false
        }
        if (password.isEmpty()) {
            toast("密码不能为空")
            return false
        }
        if (captcha.isEmpty()) {
            toast("验证码不能为空")
            return false
        }
        return true
    }

    /**
     * 两步验证，对话框
     */
    @SuppressLint("InflateParams")
    fun showTwoStepDialog(activity: Activity) {
        val dialogEt = LayoutInflater.from(activity).inflate(R.layout.dialog_et, null)
        val etCode = dialogEt.findViewById<EditText>(R.id.et_two_step_code)
        AlertDialog.Builder(activity, R.style.AppTheme_Simple)
            .setTitle("您需要进行两步验证")
            .setPositiveButton("验证") { _, _ ->
                finishLogin(etCode.text.toString(), activity)
            }
            .setNegativeButton("暂不登录") { _, _ ->
                HttpHelper.myCookieJar.clear()
                setLogin(false)

            }
            .setView(dialogEt).show()
    }

    /**
     * 两步验证，完成登录
     */
    private fun finishLogin(code: String, activity: Activity) {
        val twoStepUrl = "https://www.v2ex.com/2fa"
        vCall(twoStepUrl).enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                NetManager.dealError(activity)
            }

            override fun onResponse(call: Call, response: Response) {
                if (response.code == 200) {
                    val bodyStr = response.body?.string()!!
                    val once = Parser(bodyStr).getOnceNum()
                    val body: RequestBody = FormBody.Builder()
                        .add("code", code)
                        .add("once", once).build()
                    HttpHelper.OK_CLIENT.newCall(
                        Request.Builder()
                            .post(body)
                            .url(twoStepUrl)
                            .build()
                    )
                        .enqueue(object : Callback {
                            override fun onFailure(call: Call, e: IOException) {
                                NetManager.dealError(activity)
                            }

                            override fun onResponse(call: Call, response: Response) {
                                activity.runOnUiThread {
                                    if (response?.code == 302) {
                                        activity.toast("登录成功")
                                        setLogin(true)
                                        finish()
                                    } else {
                                        activity.toast("登录失败")
                                    }
                                }
                            }
                        })
                }
            }
        })
    }

}
